storage/rocksdb/rdb_locking_iter.h

```diff
@@ -0,0 +1,190 @@
+#include "./rdb_threads.h" /* for thd_get_current_thd */
+
+/* MyRocks header files */
+#include "./ha_rocksdb.h"
+
+namespace myrocks {
+
+//////////////////////////////////////////////////////////////////////////////
+// Locking iterator
+//////////////////////////////////////////////////////////////////////////////
+
+//
+// LockingIterator is an iterator that locks the rows before returning, as well
+// as scanned gaps between the rows.
+//
+//  Example:
+//    lock_iter= trx->GetLockingIterator();
+//    lock_iter->Seek('abc');
+//    lock_iter->Valid()==true && lock_iter->key() == 'bcd';
+//
+//   After the above, the returned record 'bcd' is locked by transaction trx.
+//   Also, the range between ['abc'..'bcd'] is empty and locked by trx.
+//
+//    lock_iter->Next();
+//    lock_iter->Valid()==true && lock_iter->key() == 'efg'
+//
+//   Now, the range ['bcd'.. 'efg'] (bounds incluive) is also locked, and there are no
+//   records between 'bcd'  and 'efg'.
+//
+class LockingIterator : public rocksdb::Iterator {
+
+  rocksdb::Transaction *txn_;
+  rocksdb::ColumnFamilyHandle* cfh_;
+  bool m_is_rev_cf;
+  rocksdb::ReadOptions read_opts_;
+  rocksdb::Iterator *iter_;
+  rocksdb::Status status_;
+
+  // note: an iterator that has reached EOF has status()==OK && valid_==false
+  bool  valid_;
+
+  ulonglong *lock_count_;
+ public:
+  LockingIterator(rocksdb::Transaction *txn,
+                  rocksdb::ColumnFamilyHandle *cfh,
+                  bool is_rev_cf,
+                  const rocksdb::ReadOptions& opts,
+                  ulonglong *lock_count=nullptr
+                  ) :
+    txn_(txn), cfh_(cfh), m_is_rev_cf(is_rev_cf), read_opts_(opts), iter_(nullptr),
+    status_(rocksdb::Status::InvalidArgument()), valid_(false),
+    lock_count_(lock_count) {}
+
+  ~LockingIterator() {
+    delete iter_;
+  }
+
+  virtual bool Valid() const override { return valid_; }
+
+  // Note: MyRocks doesn't ever call these:
+  virtual void SeekToFirst() override;
+  virtual void SeekToLast() override;
+
+  virtual void Seek(const rocksdb::Slice& target) override;
+
+  // Position at the last key in the source that at or before target.
+  // The iterator is Valid() after this call iff the source contains
+  // an entry that comes at or before target.
+  virtual void SeekForPrev(const rocksdb::Slice& target) override;
+
+  virtual void Next() override;
+  virtual void Prev() override;
+
+  virtual rocksdb::Slice key() const override {
+    assert(Valid());
+    return iter_->key();
+  }
+
+  virtual rocksdb::Slice value() const override {
+    assert(Valid());
+    return iter_->value();
+  }
+
+  virtual rocksdb::Status status() const override {
+    return status_;
+  }
+
+ private:
+  template <bool forward> void Scan(const rocksdb::Slice& target,
+                                    bool call_next) {
+    if (!iter_->Valid()) {
+      status_ = iter_->status();
+      valid_ = false;
+      return;
+    }
+
+    while (1) {
+      /*
```

Comment:
Could you thd->killed in this loop and terminate/return HA_ERR_QUERY_INTERRUPTED if set?